### PR TITLE
Added GetArtifacts API call

### PR DESCRIPTION
**Pull request by packapotatoes**
Instruction to run are in scripts/api.py

### CODE REVIEW

**Comment by vitiral** [comment]
You made these changes from master but asked to merge into branch server, which doesn't work...

... honestly, let's just go off of the master branch. I never liked the server branch anyway.

File: Cargo.toml [hunk 1]
```diff
@@ -33,6 +34,14 @@ default-features = false
 features = ["suggestions", "color", "wrap_help"]
 version = "2"
 
+[dependencies.diesel]
```
Comment (vitiral) [inline Cargo.toml]: these need to be added to the `server` feature below.
```
server = [... "diesel", "diesel_codegen"]
```

File: src/api/mod.rs [hunk 2]
```diff
@@ -49,7 +49,36 @@ fn config_json_res(res: &mut Response) {
     res.set(MediaType::Json);
     res.set(StatusCode::Ok);
 }
+/*
+fn handle_db<'a>(req: &mut Request, mut res: Response<'a>) -> MiddlewareResult<'a> {
```
Comment (vitiral) [inline src/api/mod.rs]: can you explain what this is? In general, try to remove commented code (or at least give a reason why it should stay

File: src/schema.rs [hunk 3]
```diff
@@ -0,0 +1 @@
+infer_schema!("dotenv:DATABASE_URL");
```
Comment (vitiral) [inline src/schema.rs]: I think there is another way to do this that is probably better. The way this currently works is by reading your *local running database* and using that to *compile this application*. That is a dependency you do not want for an actual product...

There should be some way to specify your scheme through sql files -- please use that instead

Comment (jmcquown) [inline src/schema.rs]: Should be fixed in my latest commit, 6a8747c

**Comment by vitiral** [comment]
looks like travis is hosed, I'll see what I can do to fix

**Comment by vitiral** [comment]
when I compile locally I get like 18 compile errors. Are you building with "--features server"?

```
# cargo build --features server
error[E0277]: the trait bound `std::string::String: diesel::types::NotNull` is not satisfied
 --> src/models.rs:3:1
  |
3 | #[derive(Insertable)]
  | ^^^^^^^^^^^^^^^^^^^^^ the trait `diesel::types::NotNull` is not implemented for `std::string::String`
  |
  = note: required because of the requirements on the impl of `diesel::types::IntoNullable` for `std::string::String`
  = note: this error originates in a macro outside of the current crate

error[E0277]: the trait bound `std::string::String: diesel::Expression` is not satisfied
 --> src/models.rs:3:1
  |
3 | #[derive(Insertable)]
  | ^^^^^^^^^^^^^^^^^^^^^ the trait `diesel::Expression` is not implemented for `std::string::String`
  |
  = note: required because of the requirements on the impl of `diesel::Expression` for `&'insert std::string::String`
  = note: required because of the requirements on the impl of `diesel::expression::AsExpression<diesel::types::Nullable<std::string::String>>` for `&'insert std::string::String`
  = note: this error originates in a macro outside of the current crate

error[E0277]: the trait bound `&'insert models::NewTestName: diesel::Insertable<schema::test_name::table, DB>` is not satisfied
 --> src/models.rs:3:1
  |
3 | #[derive(Insertable)]
  | ^^^^^^^^^^^^^^^^^^^^^ the trait `diesel::Insertable<schema::test_name::table, DB>` is not implemented for `&'insert models::NewTestName`
  |
  = help: the following implementations were found:
            <&'insert models::NewTestName as diesel::Insertable<schema::test_name::table, DB>>
  = note: required by `diesel::Insertable`
  = note: this error originates in a macro outside of the current crate

...

```

**Comment by packapotatoes** [comment]
image
image
It both compiles and runs for me.

File: src/api/mod.rs (hunk 2 above)
Comment (vitiral) [inline src/api/mod.rs]: well, you did a good job with nickel -- what you did is pretty much correct (not sure it is *performant* but we can worry about that later 😄).

I was just wondering what this was for. You can probably safely delete it.

**Comment by vitiral** [comment]
@packapotatoes I think you want to use diesel's types: http://docs.diesel.rs/diesel/types/index.html

**Comment by vitiral** [comment]
@packapotatoes make sure yours matches:
```
# rustc --version
rustc 1.16.0 (30cf806ef 2017-03-10)
```

if it doesn't:
```
rustup update
```

**Comment by vitiral** [comment]
if you're still not seeing the errors try:
```
cargo clean
```
and then run again

**Comment by vitiral** [comment]
also, make sure you are using stable: `rustup default stable`

File: Cargo.toml (hunk 1 above)
Comment (packapotatoes) [inline Cargo.toml]: Does `optional = true` for the server dependencies means these dependencies will only compile for `--features server`? Should the diesel dependencies also use this?

**Comment by vitiral** [comment]
also, you need to make sure to commit your `Cargo.lock` file, otherwise there is no guarantee that we are compiling with the same versions of libraries.

File: Cargo.toml (hunk 1 above)
Comment (vitiral) [inline Cargo.toml]: yup!

**Comment by packapotatoes** [comment]
I'm getting those errors only when I try to compile without `infer_schema!`. With `infer_schema` I have no problems.

**Comment by vitiral** [comment]
that's because `infer_schema` automatically uses diesel's types. You just need to use diesel's types for your `table!` macro.

Don't use `infer_schema!`. I honestly can't believe it exists to begin with...

**Comment by packapotatoes** [comment]
I just figured it out by using `diesel print-schema`. Trying to push the change, but I think git is out to get me.

File: src/schema.rs (hunk 3 above)
Comment (packapotatoes) [inline src/schema.rs]: To view the schema of the current database (what infer_schema! would create) run `diesel print-schema`. I used the result of that command to fix it. Hopefully this schema works for others as well.

**Comment by vitiral** [comment]
So, you should do a rebase, not a merge. Can you do the following:

```
git checkout --hard 6a8747c
git push origin server-inprogress -f
git checkout --hard e3c8aba
git fetch
git rebase origin/server-inprogress
git push origin server-inprogress
```

That will delete any local work, so make sure to stash it or save it some other way.

The normal workflow for pulling down changes would be:
```
git fetch
git rebase origin/server-inprogress
```
then you can push your own local changes.

Similarily, when pulling changes from master it is just:
```
git fetch
git rebase origin/master
```

**Comment by vitiral** [comment]
@jmcquown pushed a change, which is how things got more complicated.

**Comment by vitiral** [comment]
the other option (which you may like better) is to just start a new branch from master and manually copy+paste the stuff from here that works. It's amazing how often I have to do that when branches get borked, and in the scheme of things it really doesn't take that long.

**Comment by packapotatoes** [comment]
I went with the second option, the first way...I have no idea what went wrong. Just issued a new PR from a new branch.

**Comment by vitiral** [comment]
closing in favor of the other review

**Pull request closed without merging**